foundationdb/src/hca.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018 foundationdb-rs developers, https://github.com/bluejekyll/foundationdb-rs/graphs/contributors
+// Copyright 2013-2018 Apple, Inc and the FoundationDB project authors.
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Docs!
+//!
+
+use std::sync::Mutex;
+
+use byteorder::ByteOrder;
+use futures::future::Future;
+use futures::stream::Stream;
+use rand::Rng;
+
+use error::Error;
+use options::{ConflictRangeType, MutationType, TransactionOption};
+use subspace::Subspace;
+use transaction::{RangeOptionBuilder, Transaction};
+use tuple::Element;
+
+lazy_static! {
+  static ref LOCK: Mutex<i32> = Mutex::new(0);
+}
+
+const ONE_BYTES: &[u8] = &[0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
+
+/// High Contention Allocator
+///
+#[derive(Debug, Clone)]
+pub struct HighContentionAllocator {
+    counters: Subspace,
+    recent: Subspace,
+}
+
+impl HighContentionAllocator {
+    fn new(subspace: Subspace) -> HighContentionAllocator {
+        HighContentionAllocator {
+            counters: subspace.subspace(0),
+            recent: subspace.subspace(1),
+        }
+    }
+
+    fn allocate(
+        &self,
+        transaction: &mut Transaction,
+        subspace: &Subspace,
+    ) -> Result<Subspace, Error> {
+        loop {
+            let range_option = RangeOptionBuilder::from(self.counters.range())
```

Comment:
one question I had: this `RangeOption` gets remade a bunch of times, what's a good way to dedupe that work? Should it be set as a `const` and update `get_ranges` to work over references?

Reply:
Generally speaking, it might be easier to create a RangeOptionRef type then trying to make the existing one generic over both owned and unowned. I’ll need to refresh my memory though and review the code. I’ll try and do that ASAP.

Reply:
Reviewing the code, this would be a good change, to make a RangeOptionRef that could be constructed directly from the two vecs from get_ranges(). I'm not sure why that is returning Vecs. This might end up being a rather large and deep change though.

Reply:
One additional tricky piece about this method I later learned:  `RangeOptionBuilder::from(subspace.range())` compiles, but since `range()` returns `(Vec<u8>, Vec<u8>)`, and because we have tuple definitions for pairs + `Vec<u8>` Elements, it actually gets interpreted as a Tuple. This then produces a Range that spans `[(begin, end, 0x00), (begin, end, 0xFF))`.

Maybe we could add a key/range-convertible trait that both tuples / subspaces implement that returns `(KeySelector, KeySelector)`

Reply:
For either case, I might hold off on these changes to keep this PR scoped down

Reply:
Sorry for the late reply. Yes, I'm sure this could be cleaned up quite a bit.